packages/puppeteer-core/src/injected/PQuerySelector.ts

```diff
@@ -0,0 +1,166 @@
+import type {AwaitableIterable} from '../common/types.js';
+import {AsyncIterableUtil} from '../util/AsyncIterableUtil.js';
+import {ariaQuerySelectorAll} from './ARIAQuerySelector.js';
+import {customQuerySelectors} from './CustomQuerySelector.js';
+import {parsePSelectors, PSelector} from './PSelectorParser.js';
+import {textQuerySelectorAll} from './TextQuerySelector.js';
+import {deepChildren, deepDescendents} from './util.js';
+import {xpathQuerySelectorAll} from './XPathQuerySelector.js';
+
+class PQueryEngine {
+  #deepShadowSelectors: PSelector[][][];
+  #shadowSelectors: PSelector[][];
+  #selectors: PSelector[];
+  #selector: PSelector | undefined;
+
+  elements: AwaitableIterable<Node>;
+
+  constructor(element: Node, selector: string) {
+    selector = selector.trim();
+    if (selector.length === 0) {
+      throw new Error('The provided selector is empty.');
+    }
+    this.#deepShadowSelectors = parsePSelectors(selector);
+
+    // If there are any empty elements, then this implies the selector has
+    // contiguous combinators (e.g. `>>> >>>>`) or starts/ends with one which we
+    // treat as illegal, similar to existing behavior.
+    if (
```

Comment:
I think this should live inside the parser itself.

Reply:
The Parser doesn't check for semantically accurate selectors, just syntax.